TESTS/usb_device/basic/main.cpp

```diff
@@ -29,23 +29,205 @@
+        TEST_ASSERT_EQUAL_STRING("pass", _key);
+    }
+}
+
+void control_stress_test()
+{
+    uint16_t vendor_id = 0x0d28;
+    uint16_t product_id = 0x0205;
+    uint16_t product_release = 0x0001;
+    char _key[11] = {};
+    char _value[128] = {};
+
+    {
+        USBTester serial(vendor_id, product_id, product_release, true);
+        greentea_send_kv("control_stress_test", serial.get_serial_desc_string());
+        // Wait for host before terminating
+        greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+        TEST_ASSERT_EQUAL_STRING("pass", _key);
+    }
+}
+
+void device_reset_test()
+{
+    uint16_t vendor_id = 0x0d28;
+    uint16_t product_id = 0x0205;
+    uint16_t product_release = 0x0001;
+    char _key[11] = {};
+    char _value[128] = {};
+
+    {
+        USBTester serial(vendor_id, product_id, product_release, true);
+
+        greentea_send_kv("device_reset_test", serial.get_serial_desc_string());
+        serial.clear_reset_count();
+        // Wait for host before terminating
+        while(serial.get_reset_count() == 0);
+        greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+        TEST_ASSERT_EQUAL_STRING("pass", _key);
+
+        while(!serial.configured());
+
+        greentea_send_kv("device_reset_test", serial.get_serial_desc_string());
+        serial.clear_reset_count();
+        // Wait for host before terminating
+        while(serial.get_reset_count() == 0);
+        greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+        TEST_ASSERT_EQUAL_STRING("pass", _key);
+
+        while(!serial.configured());
+
+        greentea_send_kv("device_reset_test", serial.get_serial_desc_string());
+        serial.clear_reset_count();
+        // Wait for host before terminating
+        while(serial.get_reset_count() == 0);
+        greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+        TEST_ASSERT_EQUAL_STRING("pass", _key);
+
+        while(!serial.configured());
+
+        greentea_send_kv("device_reset_test", serial.get_serial_desc_string());
+        // Wait for host before terminating
+        greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+        TEST_ASSERT_EQUAL_STRING("pass", _key);
+    }
+}
+
+
+void device_soft_reconnection_test()
+{
+    uint16_t vendor_id = 0x0d28;
+    uint16_t product_id = 0x0205;
+    uint16_t product_release = 0x0001;
+    char _key[11] = {};
+    char _value[128] = {};
+    const uint32_t reconnect_try_count = 3;
+
+    {
+        USBTester serial(vendor_id, product_id, product_release, true);
+
+        greentea_send_kv("device_soft_reconnection_test", serial.get_serial_desc_string());
+        // Wait for host before terminating
+        greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+        TEST_ASSERT_EQUAL_STRING("pass", _key);
+
+        for(int i = 0; i < reconnect_try_count; i++) {
+            serial.disconnect();
```

Comment:
If disconnect() + complete() occur too fast the reset event will be dropped. At a minimum there should be a 200us delay between disconnect and connect. To be on the safe side I would recommend a 1ms delay, so the host controller has an entire USB frame to detect the disconnect.

Reply:
done